### PR TITLE
Adding support objects for integrating dynamic client the kubectl builder

### PR DESCRIPTION
**Pull request by krousey**
Kubectl will try to decode into `runtime.VersionedObjects`, so the `UnstructuredJSONScheme` needs to handle that intelligently.

Kubectl's builder also needs a `meta.RESTMapper` and `runtime.Typer`. The `meta.RESTMapper` requires a `runtime.ObjectConvertor` (spelling?) that works with `runtime.Unstructured`. The mapper and typer required discovery info, so I just put that in the kubectl util package since it didn't really seem to fit anywhere else. 

Subsequent PRs will be using these in kubectl.

cc @kubernetes/sig-api-machinery @smarterclayton @liggitt @lavalamp

### CODE REVIEW

**Comment by smarterclayton** [comment]
On my list today.

On May 12, 2016, at 1:42 PM, k8s-merge-robot <notifications@github.com>
wrote:

Labelling this PR as size/L

—
You are receiving this because you were assigned.
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/25540#issuecomment-218831533>

**Comment by krousey** [comment]
@k8s-bot unit test this issue: #IGNORE

File: pkg/runtime/unstructured.go [hunk 1]
```diff
@@ -146,3 +148,39 @@ func (s unstructuredJSONScheme) decodeToList(data []byte, list *UnstructuredList
...
+	unstructOut, ok := out.(*Unstructured)
+	if !ok {
+		return fmt.Errorf("output type %T in not valid for unstructured conversion")
+	}
+
+	// maybe deep copy the map? It is documented in the
+	// ObjectConverter interface that this function is not
+	// guaranteeed to not mutate the input. Or maybe set the input
+	// object to nil.
+	unstructOut.Object = unstructIn.Object
```
Comment (smarterclayton) [inline pkg/runtime/unstructured.go]: For now this is fine, I have a todo to make the conversion guarantees much more explicit, and I'll handle the implications if any here.

**Comment by smarterclayton** [comment]
Changes LGTM - although I think we want to in the future move thirdpartydata to using the unstructured codec and convertor for API calls to bypass some of the weirdness (since the two are now doing roughly similar things).

**Comment by smarterclayton** [comment]
Boilerplate failed

**Comment by krousey** [comment]
@smarterclayton thanks. Fixed.

**Comment by krousey** [comment]
or not... investigating what happened.

**Comment by krousey** [comment]
Ahh you changed ObjectKind's signature. Fix incoming shortly.

File: pkg/runtime/unstructured.go [hunk 2]
```diff
@@ -139,3 +148,39 @@ func (s unstructuredJSONScheme) decodeToList(data []byte, list *UnstructuredList
...
+	return nil
+}
+
+func (UnstructuredObjectConverter) ConvertToVersion(in Object, outVersion unversioned.GroupVersion) (Object, error) {
+	if gvk := in.GetObjectKind().GroupVersionKind(); gvk.GroupVersion() != outVersion {
```
Comment (krousey) [inline pkg/runtime/unstructured.go]: This line changed in the rebase.

File: pkg/kubectl/cmd/util/dynamic_util.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+// ObjectKind returns the group,version,kind of the provided object,
+// or an error if the object in not *runtime.Unstructured or has no
+// group,version,kind information.
+func (dot *DynamicObjectTyper) ObjectKind(obj runtime.Object) (unversioned.GroupVersionKind, error) {
+	if _, ok := obj.(*runtime.Unstructured); !ok {
+		return unversioned.GroupVersionKind{}, fmt.Errorf("type %T is invalid for dynamic object typer")
+	}
+
+	return obj.GetObjectKind().GroupVersionKind(), nil
```
Comment (krousey) [inline pkg/kubectl/cmd/util/dynamic_util.go]: This changed in the rebase.

**Comment by krousey** [comment]
Sorry I got sidetracked. I rebased this and it had to change to fit with the allocation refactorings you did. I commented on the 2 spots that changed.

**Comment by krousey** [comment]
Fixed govet errors.

**Comment by krousey** [comment]
@k8s-bot e2e test this issue:#IGNORE

couldn't pull kube-build

**Comment by krousey** [comment]
@k8s-bot test this issues: #25691

**Comment by krousey** [comment]
@k8s-bot test this issue: #25691

**Comment by krousey** [comment]
@smarterclayton is this good?

File: pkg/kubectl/cmd/util/dynamic_util.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
+/*
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/util/dynamic_util.go]: Minor nit - could you move this to `pkg/kubectl/dynamic` instead?

**Comment by krousey** [comment]
@smarterclayton nit addressed.

**Comment by krousey** [comment]
@smarterclayton is this LGTM?

**Pull request merged**